Bug Fix - Best Gift Year Rollups

dvenor opened the pull request:
# Critical Changes

# Changes

# Issues Closed
fixes #3640 
# New Metadata

# Deleted Metadata

dvenor commented:
**lurch: attach W-028500

davisagli commented:
@dvenor the "Issues Closed" section of the pull request should link to the github issue (https://github.com/SalesforceFoundation/Cumulus/issues/3640) rather than our internal work item -- our automated tooling uses this when generating release notes and will also add a comment on the issue once the fix is included in a release.

dvenor commented:
Ah.  Got it.  Thanks @davisagli.  Didn't notice the github issue.

force2b commented:
@dvenor and @davisagli I'm pretty sure it needs to be "fixes #" followed by the issue number. More info on the autoclose keywords is here: https://help.github.com/articles/closing-issues-using-keywords/

force2b commented:
@dvenor Also, in this particular case I don't think it's necessary to specify the details of what was fixed in the Changes section since the Release Notes will include a link to the fixed issue.

force2b commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+     */
+    public static String englishAlphabetOnly(String target) {
+
+            if (target == null) {
```
It looks like the indentation is off on the methods in this class.

force2b commented on src/classes/UTIL_String.cls-meta.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="urn:metadata.tooling.soap.sforce.com" fqn="UTIL_String">
```
I've never noticed an `<apexClass> tag with this format namespace string and and `fqn`. None of our other classes have that. It appears to be referencing the tooling api instead of the soap api. Honestly not sure what to make of this. @davisagli and @cdcarter What do you think about this?

dvenor replied:
I initially deployed this using VS Code and SFDX: Create Apex Class function.  Seems like it used the tooling API.  Not sure why but I'll check on it.

davisagli replied:
https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/reference_objects_overview.htm defines the `metadata.tooling.soap.sforce.com` namespace as "Objects and types that occur in both the Metadata API WSDL and the Tooling API WSDL." Given that, I think the platform must be treating them similarly. CumulusCI's code to process -meta.xml files is careful to not assume a particular xml namespace and this appears to have deployed successfully for CI via the metadata API, so I don't think it matters too much which one is used.

I assume `fqn` stands for fully-qualified name but I can't find any documentation of what it is used for here!

force2b replied:
Thanks @davisagli. I guess we can leave it. I suppose in the end, it's how the file exists in the dev/packaging org that matters, not how the tooling or metadata/soap api retrieves and formats the meta-xml file. Since the file deploys correctly and to the best of my knowledge there is nothing in the Salesforce UI that allows that part of the meta-xml file to be viewed or modified, I'm cool with it.

@dvenor Maybe you're just the first dev to create & retrieve a new class using VSCode.

lmeerkatz commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+* @description Methods used to manipulate string objects
+*
+*/
+public with sharing class UTIL_String {
```
Should this be `without sharing`? It shouldn't affect behavior either way, it just seems odd to enforce sharing on a utility class.

force2b replied:
True. It's really a candidate for `inherited sharing`, but that's not available in 43. How about no sharing declaration at all so it implicitly inherits?

dvenor replied:
Good point.  I'd think a utility class would not enforce sharing.

lmeerkatz commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+    
+    /*********************************************************************************************************
+     * @description Removes from a string all values that are not an english language alpha character
+     * @param target The the target string for the operation
```
extra 'the' :)

force2b replied:
You have much better eyes than I do! :)

lmeerkatz commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+     * @param target The the target string for the operation
+     * @return String
+     */
+    public static String englishAlphabetOnly(String target) {
```
A name with a verb might be clearer, like removeNonEnglishCharacters(String target). (Applies to multiple method names)

lmeerkatz replied:
Though I suppose that specific option could imply that you're changing the target vs. returning a stripped down value.

dvenor replied:
I originally had a verb like name but changed it.  It seemed awkward to me for some reason but I now agree it would be clearer.

lmeerkatz commented on src/classes/UTIL_String_TEST.cls:
```diff
@@ -0,0 +1,78 @@
+    /*******************************************************************************************************************
+    * @description Verifies that only english language characters are returned after processing the string
+    */
+    @isTest private static void nonEnglishAlphabetCharactersShouldBeRemoved() {
```
It would be good to include a test for characters with accents.

force2b replied:
good one!

dvenor commented on src/classes/CRLP_FiscalYears_TEST.cls:
```diff
@@ -33,14 +33,60 @@
+    /*********************************************************************************************************
+    * @description Test to ensure fiscal year is correctly formatted as a string with no punctuation
+    */
+    @isTest static void yearStringShouldNotIncludePunctuation() {
```
For some reason I though we had it in the style guide to have the `@isTest` on the same line.  Looked and clearly that is not the case. Will fix.

dvenor commented:
Modified the UTIL_String method names to start with a verb and describe what they do.  Also fixed some formatting issues and grammar issues in the descriptions.  Ready for re-review.

force2b commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /*********************************************************************************************************
+     * @description Removes from a string all values that are not a numberic character
```
spelling error in `numeric`

force2b commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+     * @param target The target string for the operation
+     * @return String
+     */
+    public static String removeNonEnglishCharacters(String target) {
```
Does "A-Z" technically refer to "English" characters? I'm wondering if this is better worded to as `removeNonAsciiLetters` and maybe `removeNonNumericAndNonAsciiLetters` below. Or possibly `removeNonAlphabeticCharacters`, etc.

force2b commented on src/classes/UTIL_String.cls:
```diff
@@ -0,0 +1,83 @@
+     * @param target The target string for the operation
+     * @return String
+     */
+    public static String removeNonNumeric(String target) {
```
Since the other methods in this class end with `Characters`, for consistency I would rename this to `removeNonNumericCharacters`.

force2b commented on src/classes/CRLP_FiscalYears.cls:
```diff
@@ -63,6 +63,16 @@ public class CRLP_FiscalYears {
+    * @param targetDate The date for which to look up the year.
+    * @return String The calendar or fiscal year for the given Date as a string.
+    */
+    public String getYearFormattedString(Date targetDate) {
```
Since this class already has a static method for `getFiscalYear(Date targetDate)`, you could create a new static method for `getYearAsFormattedString(Date theDate)`. This way all the places that call `UTIL_String.removeNonNumeric(closeDate.addMonths(n).year().format())` could just be `CRLP_FiscalYears.getYearAsFormattedString(closeDate)`. The instance method can be kept (I believe that's allowed), though it should probably be renamed to `getYearAsFormattedString`.

dvenor replied:
Do you think it's worth making this change at this time?  It's slightly more invasive and requires a bit more code rework, and I'd need to be I understand the business rules around the dates being passed into the function.  Just want to be sure I don't break something.

dvenor replied:
The `year().format()` logic is acting upon other dates, not just the current fiscal year.  Right now we don't really have a use for new static method. Even the current static method is only used within the `CRLP_FiscalYears` class.  I can see how a method that accepts a Date and return the year as a string would be useful, but I don't think that method belongs the in the `CRLP_FiscalYears` class.  Maybe it would make sense to have it in the `UTIL_String` class?

force2b replied:
Fair enough. Let's leave as is.

davisagli commented:
@force2b @dvenor the yearStringShouldNotIncludePunctuation test failed in the offset fiscal year beta org: https://mrbelvedereci.herokuapp.com/builds/37419/tests

davisagli commented:
@force2b @dvenor the offset fiscal org has a fiscal year starting in December. So maybe the quickest fix is to use a date in December for the test instead of Jan 1.

If you want to try with the same offset fiscal year config in a scratch org, run the `config_offsetfiscal` CCI flow to set it up.
